### ISSUE TITLE
Implementing RFUAV with ADALM-Pluto SDR/HackRF-One

### ISSUE THREAD

**Issue by waseem-arshad94** [kitoweeknd/RFUAV#17]
Dear Kitoo,

I am attempting to replicate and adapt your framework using an ADALM-Pluto SDR (instead of the USRP X310). Since Pluto has lower sampling rates and bandwidth, I would greatly appreciate your guidance on a few technical points to ensure my implementation aligns with your methodology:

**Dataset –** Is it possible to access the raw IQ or spectrogram dataset used in your experiments? If sharing is not possible, could you provide detailed guidelines on how you structured your dataset collection and labeling?

**Spectrogram preprocessing –** Could you confirm the final spectrogram resolution you used (e.g., 224×224, 640×640) and whether normalization was applied globally or per sample?

**YOLO setup –** Which YOLO version and input size did you use for the detection stage, and what were the key hyperparameters (batch size, epochs, learning rate)?

**ViT classifier –** Which ViT variant did you use (e.g., ViT-L/16) and what fine-tuning strategy (frozen layers vs full fine-tuning)?

**SNR augmentation –** Did you inject AWGN in the IQ domain or spectrogram domain? What was the SNR range and step size used for augmentation?

**Evaluation protocol –** How were accuracy and detection metrics computed across different SNR bins? Did you average over multiple random noise realizations?

**Hardware adaptation –** Since Pluto provides ~61.44 MS/s with narrower bandwidth compared to the USRP X310, do you foresee any limitations or modifications necessary to adapt your framework to Pluto?

**Implementation details –** Are the training scripts, configurations, or pretrained weights for YOLO and ViT available? If not, could you kindly share your recommended hyperparameters?

Your guidance would be extremely valuable in helping me replicate your excellent work on more accessible hardware, making RFUAV-style approaches more widely usable.
This is my email: waseemarshad94@gmail.com , if it is possible to share the scripts for recording and analysis. 
Thank you for your time and for this important contribution to UAV detection research. I look forward to your response.

Regards

**Comment by kitoweeknd** [comment]
Hi @waseem-arshad94 , thanks for your questions! Here are my answers — I hope this helps.

### **Dataset**

A portion of the raw data from my dataset has been open-sourced on Hugging Face. Due to its large size, I’m unable to release the entire dataset at this time, but I plan to make it available once I find a better deployment solution. The data collection process is clearly described in my arXiv paper, so please feel free to ask if you have any questions about specific steps. I’d also like to emphasize that if you intend to collect signals yourself, the sampling rate should be at least **100 MHz**. Labeling is not particularly challenging — it can be organized in the same way as a standard classification dataset.

### **Spectrogram Preprocessing**

The resolution of the preprocessed spectrogram images varies depending on the network type. For example, ViT uses **224×224**, while other CNNs use **640×640**. This resizing is handled automatically by the network when PyTorch loads the dataset.

### **YOLO Setup**

The YOLO model we used in this project is **YOLOv5s**. You can find the hyperparameters in `utils.DetModels.yolo.cfg`. During training, we used **50 epochs** and a **batch size of 64**. By the way, both epochs and batch size seem to have only a minor effect on the results.

### **ViT Classifier**

The classifier used in our demo is **ResNet-18**, and we conducted extensive comparisons with mainstream CNNs and vision models such as ViT and MobileNet. The best-performing network was **ViT-L/16**, which we trained using full fine-tuning.

### **SNR Augmentation**

We injected **AWGN** into the IQ data. Specifically, we applied AWGN modulation on the original drone IQ signals, adjusting the SNR from **-20 dB to 20 dB** in **2 dB steps**. This results in multiple augmented datasets at different SNR levels (20 dB, 18 dB, 16 dB, …, -18 dB, -20 dB), in addition to the original data.

### **Evaluation Protocol**

We used both **accuracy (ACC)** and **overall accuracy (OA)** to evaluate model performance under different SNRs and SNR ranges, so in fact we applied both methods you mentioned.

### **Hardware Adaptation**

You raised a very good point. For RF acquisition equipment, a higher sampling rate usually means a higher cost. Devices like **Pluto SDR** can achieve similar effects, but the types of drones it can identify are more limited. In practice, using Pluto is like observing only a portion of the spectrum — this causes feature loss, making some drone types indistinguishable. If you use equipment with a lower sampling rate, retraining the model will be necessary.

### **Implementation Details**

You can find all training configurations under `RFUAV.configs`. The training scripts can be run with:

```bash
python train.py --config <your_config_file>
```

---

Feel free to ask if anything is unclear and continue posting on this issue — I’ll be happy to reply as soon as possible. Thanks again for using our dataset, and I hope this project is helpful for your research.

**Comment by waseem-arshad94** [comment]
@kitoweeknd Thank you so much for your response. I will try to reproduce your results and get back to you soon.

**Comment by elgup0** [comment]
Waseem, when the instantaneous bandwidth (IBW) is narrower, the receiver must sweep more steps to cover the same spectrum. This increases overall scan time. On top of that, inference speed also slows down because the system has to process more batches of spectrogram images, which compounds into additional time delay.

**Comment by kitoweeknd** [comment]
Hi @waseem-arshad94

I’m really glad to hear that you successfully implemented our framework and even integrated it with a Pluto SDR for real-time testing — that’s great work!

First of all, I’d like to apologize for the delayed response. My schedule has been quite hectic lately, but I’ll make sure to carefully review your code and provide a more detailed answer once I get some time.

For now, here are some quick suggestions that might help you address the issue you’re seeing (where the model consistently predicts *DAUTEL EVO Nano*):

1. **Apply additional data augmentation**, especially to the samples from drones other than *DAUTEL EVO Nano*. This helps the model generalize better across different transmitter types.
2. **Balance the dataset** during training — make sure each drone category has roughly the same number of images. Imbalanced data can cause the model to bias toward the majority class.

I’ll take a closer look at your shared files and get back to you with more detailed feedback soon.

Looking forward to your updates!

**Comment by kitoweeknd** [comment]
I'm very glad my advice was helpful. I've had some free time recently, so please feel free to ask me any questions. 👍🏻
